runtime/src/blueoil.cpp

```diff
@@ -15,6 +17,172 @@
 
 
 namespace blueoil {
+
+
+int Tensor::shapeVolume() {
+    return this->shapeVolume(m_shape);
+}
+
+int Tensor::shapeVolume(std::vector<int> shape) {
+    return std::accumulate(shape.begin(), shape.end(),
+			   1, std::multiplies<int>());
+}
+
+
+Tensor::Tensor(std::vector<int> shape)
+    : m_shape(shape),
+      m_data(std::vector<float>(this->shapeVolume(std::move(shape)), 0)) {
+}
+
+Tensor::Tensor(std::vector<int> shape, std::vector<float> data)
+    : m_shape(std::move(shape)),
+      m_data(std::move(data)) {
+}
+
+Tensor::Tensor(std::vector<int> shape, float *arr)
+    : m_shape(shape),
+      m_data(std::vector<float>(arr,
+				arr + this->shapeVolume(std::move(shape)))) {
+}
+
+Tensor::Tensor(const Tensor &tensor)
+    : m_shape(tensor.m_shape),
+      m_data(tensor.m_data) {
+}
+
+
+std::vector<int> Tensor::shape() const {
+    return m_shape;
+}
+
+std::vector<float> &Tensor::data() {
+  return m_data;
+}
+
+float *Tensor::dataAsArray() {
+  if (this->m_shape.size() == 0) {
```

Comment:
Btw, writing everything as `this->m_xxx` is not C++ style, but rather python one, where you have to reference `self` explicitly.
You already prefixed all members with `m_` so it is already clear which values are members and which are not.